### ISSUE TITLE
Stack guards not supported in this environment

### ISSUE THREAD

**Issue by cyberkaidev** [supabase/gotrue-js#746]
# Bug report

- [ ] I confirm this is a bug with Supabase, not with my own application. 
- [ ] I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com). 

## Describe the bug

@supabase/gotrue-js: Stack guards not supported in this environment. Generally not an issue but may point to a very conservative transpilation environment (use ES2017 or above) that implements async/await with generators, or this is a JavaScript engine that does not support async/await stack traces.

## To Reproduce

I was following the [doc](https://supabase.com/docs/guides/getting-started/tutorials/with-expo), I created the client and I was able to query the database, but a message keeps appearing in my app of:"@supabase/gotrue-js: Stack guards not supported in this environment. Generally not an issue but may point to a very conservative transpilation environment (use ES2017 or above) that implements async/await with generators, or this is a JavaScript engine that does not support async/await stack traces.".
I'm using expo.

## System information

- OS: macOS
- Version of supabase-js: ^2.26.0
- Version of expo: ^49.0.0

**Comment by saoud77** [comment]
Same here! I'm experiencing the same thing.

**Comment by saoud77** [comment]
> I've got the same error after upgrading my expo SDK from 48 to 49. Warning appears regardless of `@supabase/supabase-js` version. Tested and was able to reporoduce in both:
> 
> * ✅ `"@supabase/supabase-js": "^2.12.0"`
> * ✅ `"@supabase/supabase-js": "^2.26.0"`

I've tried both versions of '@supabase/supabase-js' along with Expo version 48, but the error still persists

**Comment by saoud77** [comment]
Same here! I'm experiencing the same thing.

**Comment by florianjuengermann** [comment]
Same here!

**Comment by Honest0** [comment]
Same here!
I faced same issue on my vercel environment now.

**Comment by insanrizky** [comment]
Same here!

**Comment by m1thrandir225** [comment]
Same here on Expo SDK 48 with @supabase/supabase-js "^2.24.0"

**Comment by REDJEM-Amir** [comment]
If the warning bothers you, you can remove it like this:

`const originalWarn = console.warn
console.warn = (message, ...args) => {
  if (message.startsWith('@supabase/gotrue-js')) return
  originalWarn(message, ...args)
}

import 'react-native-url-polyfill/auto'
import * as SecureStore from 'expo-secure-store'
import { createClient } from '@supabase/supabase-js';`

**Comment by masnwilliams** [comment]
It throws a full error for me and crashes my vercel deployment. I don't think the code above will help in this case.

**Comment by j4w8n** [comment]
> It throws a full error for me and crashes my vercel deployment. I don't think the code above will help in this case.

Can you share the error and any other info you believe would be relevant?

**Comment by masnwilliams** [comment]
> Can you share the error and any other info you believe would be relevant?

Of course, here is the error that is from my build logs on Vercel when trying to load my production build:

```
@supabase/gotrue-js: Stack guards not supported in this environment. Generally not an issue but may point to a very conservative transpilation environment (use ES2017 or above) that implements async/await with generators, or this is a JavaScript engine that does not support async/await stack traces. Safari is known to not support stack guards.
```

I will also note that this only appears after I authenticate as shown in this video:

https://github.com/supabase/gotrue-js/assets/43387599/2ef8185f-24b4-42ab-8be9-1ae0f1de1cfc

My entire application works perfectly on a preview build but fails once in production.

**Comment by j4w8n** [comment]
Roger. I should've been more clear. I thought Vercel would throw an additional log because you said it crashes.

**Comment by masnwilliams** [comment]
Yeah, that is the log that Vercel gives in my build log and the screenshot below is what my prod website looks like instead of loading the UI 😂

Luckily this is a brand new product and has 0 users (like many of side projects lol) but I did use nextjs/vercel/supabase for a business I had before that had a lot of users and if something like this happened, I would be rolling back and considering switching to another cloud backend service.

<img width="1683" alt="Screenshot 2023-07-27 at 4 41 52 PM" src="https://github.com/supabase/gotrue-js/assets/43387599/45cda1e9-85c8-439b-875c-17273215433b">

**Comment by j4w8n** [comment]
> This does not occur on pages where I do not call supabase.auth.getSession() from @supabase/auth-helpers-nextjs.

@masnwilliams I do see in the code where calling `getSession()` leads to a stack guard check or whatever. But there is also one when the client is initialized, so I'm not sure why it only chokes on those pages.

**Comment by masnwilliams** [comment]
@j4w8n Interesting update...

So I decided to remove my custom CI/CD GitHub Action pipeline that uses `vercel deploy --prebuilt --prod --token=TOKEN` and just run the build like a regular deploy on Vercel and everything works fine with no errors!

So I think it has to do with deploying a `prebuilt` build to vercel 👍 

Still narrowing it down and will respond again soon with final details

**Comment by masnwilliams** [comment]
Ok so I've narrowed this down to, at least in my case, the single cause.

It is the `--prod` flag on `vercel deploy --prebuilt --prod --token=TOKEN`.

Works:
- `vercel deploy --prebuilt --token=TOKEN`

Does Not Work:
- `vercel deploy --prebuilt --prod --token=TOKEN`

### Summarization

For some reason, [Vercel Deploy Prod](https://vercel.com/docs/cli/deploy#prod) breaks the stack overflow guard that supabase gotrue auth implements.

So if you are deploying a project that uses Supabase Auth to Vercel using [Vercel CLI](https://vercel.com/docs/cli), do not use the `--prod` tag. This ended up not being an issue for me because it seems like all the `--prod` tag does, at least from what I can see, is automatically set the build to direct to the production domain. Since I set my `main` repo branch to a specific domain, that solves that.

Hopefully this helps anyone who had this issue @atanaskanchev  @mallahyari @saoud77 @florianjuengermann @Honest0 @insanrizky @m1thrandir225

**Comment by m1thrandir225** [comment]
> 

What about on expo? Is this something with the babel / typescript config on expo or ?

**Comment by masnwilliams** [comment]
> What about on expo?

I'm not sure about this on expo. I am using Next.js so I can only speak for that unfortunately

**Comment by abhishiv** [comment]
This isnt an harmless error I think. It's also causing db queries to not work without any error in safari. Anyone has a idea why?

**Comment by j4w8n** [comment]
> This isnt an harmless error I think. It's also causing db queries to not work without any error in safari after you call `getSession()`. Anyone has a idea why?
> 
> This means supabase is broken on safari.

I do see some references to safari in the [helper functions](https://github.com/supabase/gotrue-js/blob/master/src/lib/helpers.ts), for code where they try to mitigate some safari issues. Maybe there are more needed??

If you can provide a minimum repo, or a code example, that might help out the maintainers.

**Comment by abhishiv** [comment]
Hey @j4w8n 

Thanks for your reply. I'm using it like this

```
  supabase.auth.onAuthStateChange(async (event, session) => {
    console.log("sessions", event, session, session && session.user);
    if (session && session.user) {
      console.log("fetching");
      const { data, error } = await supabase
        .from("user_profiles")
        .select("*")
        .eq("id", session.user.id);
      console.log(data, error);
    } 
  });
```

The db query from user_profiles is silently swallowed and it never returns. This only happens on safari(I'm using v16.3) and it works without any issues in chrome.

**Comment by masnwilliams** [comment]
## A further update to my previous replies

I now have the `--prebuilt` flag working as intended in a github workflow.

**Comment by dejongyeong** [comment]
I have been dealing with this issue as well when using the latest packages to `getSession()` in `middleware.ts` after deploying to Vercel and testing it out with Chrome and Edge.

I found out that the older versions work but not the latest ones when deploying to Vercel after hours of testing and researching the error.

I have no choice at the moment but to downgrade the following packages:

```
"@supabase/auth-helpers-nextjs": "0.7.2",
"@supabase/gotrue-js": "2.34.0",
"@supabase/supabase-js": "2.26.0",
```

I will look back into this or upgrade to the latest version once it is more stable in Vercel Production.

**Comment by virgiliud** [comment]
I'm seeing the same error on Vercel.

> @supabase/gotrue-js: Stack guards not supported in this environment. Generally not an issue but may point to a very conservative transpilation environment (use ES2017 or above) that implements async/await with generators, or this is a JavaScript engine that does not support async/await stack traces. Safari is known to not support stack guards.

**Comment by masnwilliams** [comment]
Supabase Auth team is looking into this now, hopefully we get a fix soon. It very likely could be related to using `getSession()` in middleware, which I do as well as @dejongyeong. However unlike @dejongyeong, reverting back to the version mentioned did not solve my problem.

**Comment by ryszardwhee** [comment]
We are experiencing the same issue in production.
We are hosting the code on vercel. The warning happens in middleware.ts when we call function: `supabase.auth.refreshSession`. 

`middleware.ts` hang in production and timeouts.

Packages version:
"next": "13.4.12",
"@supabase/supabase-js": "^2.29.0"

**Comment by dalkommatt** [comment]
> It throws a full error for me and crashes my vercel deployment. I don't think the code above will help in this case.

Same issue here with the same error in my Vercel logs: 
```@supabase/gotrue-js: Stack guards not supported in this environment. Generally not an issue but may point to a very conservative transpilation environment (use ES2017 or above) that implements async/await with generators, or this is a JavaScript engine that does not support async/await stack traces. Safari is known to not support stack guards.```

Using > ES2017 did not fix this issue for me.

**Comment by hf** [comment]
Hey everyone, this is a _warning._ If you see it, your app should generally work as it worked many versions ago. It does mean that some new improvements to the library will probably not be supported. Please continue to report problems with it, but it does not mean that when you see this warning the behavior is because of it.

With regards to Expo: Expo uses WebKit which does not support async/await stack traces. Generally however in Expo / React Native / other WebKit based mobile app frameworks this can safely be ignored. I'll try to find a way to detect Expo / React Native and remove the log if it's bothering you.

With regards to Vercel, just seeing this now on supabase.com. This is a recent change as I am absolutely certain that 2 weeks ago stack guards worked. 

@abhishiv Just seeing your problem. I suspect that does not work, because recently we made `onAuthStateChange` callbacks wait when notified before returning. Because stack guards are not supported on Safari / WebKit, the way you're using it may be causing async/await recursion and the callback is probably not completing. I'm trying to see if there's some other way this can be achieved to get rid of this issue. A quick fix could be, for example:

```typescript
  supabase.auth.onAuthStateChange(async (event, session) => {
    console.log("sessions", event, session, session && session.user);
    if (session && session.user) {
      console.log("fetching");
      (async () => {
        const { data, error } = await supabase
          .from("user_profiles")
          .select("*")
          .eq("id", session.user.id);
        console.log(data, error);
       })()
    } 
  });
```

This will start the query outside the callback and the recursive wait should be gone.

**Comment by abhishiv** [comment]
hey @hf thanks for your reply!

Yeah, I was able to get around it with a similar technique using setTimeout.

```
  const handler = async (event: any | undefined, session: Session | null) => {
    if (session && session.user) {
      const { data, error } = await supabase
        .from("user_profiles")
        .select("*")
        .eq("id", session.user.id);
      if (error) {
        return;
      }
      const profile = data[0];
      $session({ session, profile });
    } else {
      $session(null);
    }
  };
  onMount(async () => {
    const session = await supabase.auth.getSession();
    handler(undefined, session.data.session);
    // https://github.com/supabase/gotrue-js/issues/746#issuecomment-1655484486
    setTimeout(() => supabase.auth.onAuthStateChange(handler), 0);
  });
```

However it would be great to have this behaviour documented since because of this issue I pushed a broken version in production and then had to spend a lot of time debugging it.

**Comment by hf** [comment]
>  pushed a broken version in production and then had to spend a lot of time debugging it.

Very sorry about this! I'm trying to find a way to fix this... Safari/WebKit is proving especially difficult. Also, please remember you can always revert back to the version that worked for you previously.

**Comment by milovangudelj** [comment]
I had the same issue with Next.js `v13.4.12` and Supabase.js `v2.31.0`.

I got the error when I tried to use supabase in an edge function. 

```ts
export const runtime = 'edge'
```
Removing it solved the issue.

**Comment by iodamiano** [comment]
I have the same error, on iOS i can start the project and having only the warning, if i try to launch the Web version crash all the things.

**Comment by dmitryshostak** [comment]
I think we've just encountered the same issue. Seems like Bing/Yahoo crawlers can't render our app. 
(Nextjs, Vercel, @supabase/auth-helpers-nextjs)

**Request User Agent**: Mozilla/5.0 AppleWebKit/537.36 (KHTML, like Gecko; compatible; bingbot/2.0; +http://www.bing.com/bingbot.htm) Chrome/103.0.5060.134 Safari/537.36
**Type**: Edge Middleware

In Vercel logs also we get
`
@supabase/gotrue-js: Stack guards not supported in this environment. Generally not an issue but may point to a very conservative transpilation environment (use ES2017 or above) that implements async/await with generators, or this is a JavaScript engine that does not support async/await stack traces.
`

**Comment by milovangudelj** [comment]
I'd be happy to test it out. How can I update the `gotrue-js` version used by `supabase-js` to try out the new implementation?

**Comment by hf** [comment]
@milovangudelj Usually you can use [`npm link`](https://docs.npmjs.com/cli/v8/commands/npm-link) or equivalents.

For example:

1. Clone the repository.
2. `cd` into it.
3. `npm i` then `npm run build`
4. `npm link`
5. `cd` into your project
6. `npm link @supabase/gotrue-js` 
7. When you build your project, it should pick up the cloned repo. However, this doesn't always work well so you might have to play around a few times to get it to work. Inspect all `node_module` directories in your project to see if `gotrue-js` links back to the cloned folder.

**Comment by kamto7** [comment]
> Done that. It works on local, but I have no way of testing it on Vercel when deploying the app. There it's still going to pull the latest version and error out when deployed... Note that the error happens not during the build step but when invoking the api function containing both a Supabase client and `export const runtime = 'edge'`

Same to me. I don't know how this will affect the business.

**Comment by silentworks** [comment]
If you use [yalc](https://github.com/wclr/yalc) instead you can test this on Vercel.

For example:

- Install yalc using `npm i yalc -g`
- Clone the repository.
- cd into it.
- `npm i` then `npm run build`
- yalc publish
- cd into your project
- yalc add @supabase/gotrue-js
- modify your package.json to override the gotrue-js version
if you are using npm you will need to add:
```json
"overrides": {
	"@supabase/gotrue-js": "file:.yalc/@supabase/gotrue-js"
}
```
if pnpm:
```json
"pnpm": {
	"overrides": {
		"@supabase/gotrue-js": "file:.yalc/@supabase/gotrue-js"
	}
}
```
- When you build your project, it should pick up the cloned repo.
- You can now commit the `.yalc` directory to your project and deploy to Vercel, this way it will use the gotrue you just added into the project a build time and runtime.

**Comment by milovangudelj** [comment]
Ok, thanks! I'll try this later and I'll let you know how it goes ✌🏻

**Comment by Diegofreema** [comment]
I am having this same issue too with my expo project, it works in developement but crashes in production.
i even had to install a previous version og supabase, but it still did not work.

**Comment by makeusabrew** [comment]
I've manually bumped my `@supabase/gotrue-js` dependency to v2.47.0 until https://github.com/supabase/supabase-js/pull/835 is merged. So far, so good. See my comment in that PR thread: https://github.com/supabase/supabase-js/pull/835#issuecomment-1701340535

**Comment by picklebrownie** [comment]
I am having this same issue. when the user submits the file, the error occurs. 

code:
`'use client'

import { Button } from '@/components/Button'
import { Input } from '@/components/Input'
import { Label } from '@/components/Label'
import { toast } from '@/components/UseToast'
import { createClientComponentClient } from '@supabase/auth-helpers-nextjs'
import { useRouter } from 'next/navigation'
import React, { useRef, useState } from 'react'
import { uuid } from 'uuidv4'

export default function UploadFile() {
  const supabase = createClientComponentClient()

  const [loading, setLoading] = useState(false)

  const [selectedFile, setSelectedFile] = useState<File | null>(null)
  const [fileValue, setFileValue] = useState<string>('')
  const [fileName, setFileName] = useState<string>('')

  const [fileAlias, setFileAlias] = useState<string>('')

  // Ref object to reference the input element
  const inputFile = useRef<HTMLInputElement>(null)

  // Handle file upload event
  async function uploadFile(event: React.FormEvent<HTMLFormElement>) {
    setLoading(true)
    event.preventDefault()

    const bucket = 'files'

    // check that a file exists
    if (!selectedFile) {
      // e.target.value = null
      setLoading(false)
      toast({
        variant: 'destructive',
        title: 'Oops...',
        description: 'No file found! Please try again.',
      })
      return
    }

    // get supabase user id
    const {
      data: { session },
    } = await supabase.auth.getSession()
    const userId = session?.user.id

    // add file to storage bucket

    const file_url = `${userId}/${uuid()}-${selectedFile.name}`

    // Call Storage API to upload file
    try {
      const { data: data2, error: error2 } = await supabase.storage
        .from(bucket)
        .upload(file_url, selectedFile)

      // Handle error if upload failed
      if (error2) {
        // console.log(
        //   'Documents > UploadFile > upload file error ',
        //   error2,
        // )
        // e.target.value = null
        toast({
          variant: 'destructive',
          title: 'Oops...',
          description: 'Error uploading the file! Please try again.',
        })
        setLoading(false)
        return
      }
    } catch (error) {
      toast({
        variant: 'destructive',
        title: 'Oops...',
        description: 'Error uploading the file! Please try again.',
      })
      setLoading(false)
      return
    }

    // add entry to user_files table
    const { data, error } = await supabase.from('user_files').insert({
      file_url: file_url,
      file_name: fileAlias,
      user_id: userId,
    })

    // Handle error if upload failed
    if (error) {
      // console.log(
      //   'Documents > UploadFile > upload record error ',
      //   error,
      // )
      // e.target.value = null
      toast({
        variant: 'destructive',
        title: 'Oops...',
        description: 'Error uploading the file! Please try again.',
      })
      setLoading(false)
      return
    }

    // redirect to dashboard/documents

    // handle what happens after the file is uploaded
    // event.target.value = null
    toast({
      title: 'Success!',
      description: 'File uploaded successfully!',
    })
    setFileAlias('')
    handleReset(event)
    setLoading(false)
  }

  const fileChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    if (event.target.files && event.target.files.length > 0) {
      setSelectedFile(event.target.files[0])
    }
  }

  // Function to reset the input element
  const handleReset = (event: any) => {
    event.preventDefault()
    setSelectedFile(null)
    if (inputFile.current) {
      inputFile.current.value = ''
      inputFile.current.type = 'text'
      inputFile.current.type = 'file'
    }
  }
  return (
    <div>
      <div>
        <form
          onSubmit={uploadFile}
          id="form"
        >
          <div>
            <div className="mt-10 space-y-8 border-b border-gray-900/10 pb-12 sm:space-y-0 sm:divide-y sm:divide-gray-900/10 sm:border-t sm:pb-0">
              <div className="sm:grid sm:grid-cols-3 sm:items-start sm:gap-4 sm:py-6">
                <Label htmlFor="alias">File Name</Label>
                <div className="mt-2 sm:col-span-2 sm:mt-0">
                  <Input
                    type="text"
                    name="alias"
                    value={fileAlias}
                    required
                    disabled={loading}
                    onChange={(e) => setFileAlias(e.target.value)}
                    placeholder="Appointment Bill 08/13/2023"
                    className="w-full sm:max-w-sm"
                  />
                </div>
              </div>
              <div className="sm:grid sm:grid-cols-3 sm:items-start sm:gap-4 sm:py-6">
                <Label htmlFor="file">File</Label>
                <div className="mt-2 sm:col-span-2 sm:mt-0 flex flex-row">
                  <Input
                    type="file"
                    name="file"
                    ref={inputFile}
                    disabled={loading}
                    onChange={fileChange}
                    className="block w-full sm:max-w-sm"
                  />
                  {selectedFile && (
                    <Button
                      disabled={loading}
                      onClick={handleReset}
                      variant="secondary"
                      className="ml-4"
                    >
                      {loading ? 'Uploading...' : 'Remove File'}
                    </Button>
                  )}
                </div>
              </div>
            </div>
          </div>
          <div className="mt-6 flex items-center justify-end gap-x-6">
            <Button
              disabled={loading}
              type="submit"
            >
              {loading ? 'Uploading...' : 'Upload File'}
            </Button>
          </div>
        </form>
      </div>
    </div>
  )
}
`

error on vercel:
`
Request User Agent
Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Mobile Safari/537.36
Level
Error

Environment
production
Branch
main
Type
Edge Middleware

Function
/middleware
@supabase/gotrue-js: Stack guards not supported in this environment. Generally not an issue but may point to a very conservative transpilation environment (use ES2017 or above) that implements async/await with generators, or this is a JavaScript engine that does not support async/await stack traces. Safari is known to not support stack guards.

Dashboard – Vercel

`

any help would be appreciated. this prevents my users from making uploads on mobile.

**Comment by zohaib304** [comment]
I've updated to version @supabase/gotrue-js 2.51.0, and the issue is now resolved.

**Comment by j4w8n** [comment]
> I've updated to version @supabase/gotrue-js 2.51.0, and the issue is now resolved. 

Yeah I believe they went with a different mechanism for "locking", in a newer version.

**Comment by 0ax1** [comment]
I think this was resolved as part of https://github.com/supabase/gotrue-js/pull/757. @cyberkaidev Can this issue be closed?

**Comment by amzamani** [comment]
Same here, I am not getting the error in local development, but in the vercel logs of my serverless application
